execution/src/gas.rs

```diff
@@ -205,6 +235,83 @@ impl L1GasFee for EcotoneGasFee {
     }
 }
 
+#[cfg(feature = "op-upgrade")]
+#[derive(Debug)]
+pub struct FjordGasFee {
+    base_fee: U256,
+    base_fee_scalar: U256,
+    blob_base_fee: U256,
+    blob_base_fee_scalar: U256,
+}
+
+#[cfg(feature = "op-upgrade")]
+impl FjordGasFee {
+    const GAS_PRICE_MULTIPLIER: U256 = U256::from_limbs([16, 0, 0, 0]);
+    /// Absolute part of the negative intercept
+    const INTERCEPT_ABS: u32 = 42_585_600;
+    const FAST_LZ_COEF: u32 = 836_500;
+    const MIN_TX_SIZE: u32 = 100;
+
+    pub fn new(
+        base_fee: U256,
+        base_fee_scalar: u32,
+        blob_base_fee: U256,
+        blob_base_fee_scalar: u32,
+    ) -> Self {
+        Self {
+            base_fee,
+            base_fee_scalar: U256::from(base_fee_scalar),
+            blob_base_fee,
+            blob_base_fee_scalar: U256::from(blob_base_fee_scalar),
+        }
+    }
+}
+
+#[cfg(feature = "op-upgrade")]
+impl L1GasFee for FjordGasFee {
+    fn l1_fee(&self, input: L1GasFeeInput) -> U256 {
+        use umi_shared::primitives::I256;
+
+        // The spec <https://specs.optimism.io/protocol/fjord/exec-engine.html#fjord-l1-cost-fee-changes-fastlz-estimator>
+        // returns a `U256` as the final result, so we can widen the types in advance while
+        // adding the intercept sign (due to clunkiness of converting from non-ruint types).
+        let intercept = -I256::from_raw(U256::from(Self::INTERCEPT_ABS));
+        let min_tx_size = U256::from(Self::MIN_TX_SIZE);
+        let fast_lz_coef = U256::from(Self::FAST_LZ_COEF);
+        let fast_lz_size = input.fast_lz_size;
+
+        let estimated_size_scaled = {
+            let scaled = I256::from(fast_lz_coef * fast_lz_size) + intercept;
+            let min_scaled = I256::from(min_tx_size * U256::from(1_000_000));
+            U256::from(scaled.max(min_scaled))
+        };
+
+        let weighted_gas_price = Self::GAS_PRICE_MULTIPLIER * self.base_fee_scalar * self.base_fee
+            + self.blob_base_fee_scalar * self.blob_base_fee;
+
+        // We scale down by 1e6 instead of 1e12 to preserve the previous Ecotone omission of
```

Comment:
I am not sure why we skipped the 1e6 factor originally. This is something we will maybe need to review again in the future.